racket/src/io/network/tcp-port.rkt

```diff
@@ -92,3 +95,52 @@
     [(tcp-output-port? cp)
      (set-tcp-output-port-abandon?! cp #t)
      (close-port p)]))
+
+(define max-file-offset-for-platform
+  (case (system-type 'word)
+    [(32) (sub1 (expt 2 31))]
+    [(64) (sub1 (expt 2 63))]))
+
+(define/who (tcp-copy-file dst-p path
+                           [start 0]
+                           [end (file-size path)])
+  (define op (->core-output-port dst-p))
+  (check who tcp-output-port? op)
+  (check who path-string? path)
+  (check who exact-nonnegative-integer? start)
+  (check who exact-positive-integer? end)
+  (unless (< start max-file-offset-for-platform)
+    (raise-argument-error who (string-append "(</c #x" (number->string max-file-offset-for-platform 16) ")") 2 dst-p path start end))
+  ;; TODO(bogdan): should this be relaxed to (<= end (+ max-... chunksize))?
+  (unless (<= end max-file-offset-for-platform)
+    (raise-argument-error who (string-append "(<=/c #x" (number->string max-file-offset-for-platform 16) ")") 3 dst-p path start end))
+  (unless (< start end)
+    (raise-arguments-error who "start must be less than end" "start" start "end" end))
+
+  (define size (file-size path))
+  (unless (<= end size)
+    (raise-argument-error who (string-append "(<=/c " (number->string size) ")") 3 dst-p path start end))
+
+  (define chunksize #x7FFFFFFD)
+  (define ip (->core-input-port (open-input-file path)))
+  (define dst-fd (fd-output-port-fd op))
+  (define src-fd (fd-input-port-fd ip))
+  (define status (rktio_make_sendfile_status rktio))
+  (let loop ([offset start]
+             [nbytes (- end start)])
+    (define sent
+      (rktio_sendfile rktio dst-fd src-fd offset (min nbytes chunksize) status))
+    (cond
+      [(rktio-error? sent)
+       (rktio_sendfile_status_free rktio status)
+       (close-input-port ip)
+       (send tcp-output-port op raise-write-error sent)]
+
+      [(zero? (- nbytes sent))
+       (rktio_sendfile_status_free rktio status)
+       (close-input-port ip)]
+
+      [else
+       ;; TODO(bogdan): Need to allow breaks?
+       (sync op)
```

Comment:
Is it OK to do something like `((if (break-enabled) sync/enable-break sync) op)` at this level? Can I then wrap that in an exception handler to free resources when a break is raised? I think the answer to both of those is "yes", but I figured it'd be best to ask because I haven't noticed any other code in `src/io` do that.

Reply:
It's not ok to allocate a primitive resource, like `rktio_make_sendfile_status`, and then allow breaking without freeing the resource. Probably a sendfile-status obbject will need to be registered with the current custodian.

More generally, having a `sync` in the middle of a function suggests that it's not the right core function, and that function producing an evt is the right idea instead.